### PR TITLE
Teach CircleCI to run eslint for files modified in the topic branch

### PR DESCRIPTION
**Pull request by nssw**
This PR seeks to teach CircleCI to run-lint for the files modified in the topic branch. As a consequence, if we set to `errors` eslint rules which are now `warnings`, CircleCI report would only give contextual information about the PR, not the whole errors in the codebase.

### Test case

You might want to see how this works in the test case PR #6658 and what the results would be in Circle CI [here](https://circleci.com/gh/Automattic/wp-calypso/20285). The test case sets to `error` the rule `space-before-blocks` and introduces a break in the rule in file `client/components/accordion/section.jsx`. Expected result is that only that file is linted when running CircleCI although there are more cases in the codebase where that rule is not followed.

only_lints_files_in_branch

cc @nb

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -6,13 +6,7 @@ test:
     - NODE_ENV=test make client/config/index.js:
         parallel: true
   override:
-    - bin/run-lint :
-        parallel: true
```
Comment (nb) [inline circle.yml]: Are we going to lose the checks running in parallel?

File: circle.yml [hunk 2]
```diff
@@ -6,13 +6,7 @@ test:
...
-        files:
-          - client/**/*.js
-          - client/**/*.jsx
-          - server/**/*.js
-          - server/**/*.jsx
+    - bin/run-lint $(git diff --name-only $(git merge-base $(git rev-parse --abbrev-ref HEAD) origin/master)..HEAD *js *jsx)
```
Comment (nb) [inline circle.yml]: This way wouldn’t we run the checks *only* on the changed files? We still want to show problems in other files/lines, but just relegate errors to warnings.

File: package.json [hunk 3]
```diff
@@ -142,7 +142,7 @@
     "test-client:watch": "nodemon -e js,jsx --exec npm run test-client",
     "test-server:watch": "nodemon -e js,jsx --exec npm run test-server",
     "test-test:watch": "nodemon -e js,jsx --exec npm run test-test",
-    "lint": "bin/run-lint",
+    "lint": "bin/run-lint .",
```
Comment (nb) [inline package.json]: Why did you change the way we call `run-lint`? I am not saying it’s wrong, just wondering why.

File: circle.yml (hunk 1 above)
Comment (nssw) [inline circle.yml]: The [files modifier](https://circleci.com/docs/parallel-manual-setup/#using-configuration-files-modifier) accepts only globs, so we cannot use them to write git expressions. We could find a way to set parallelism up again [by using env variables](https://circleci.com/docs/parallel-manual-setup/#env-splitting) in `run-lint` script, though.

The reason I was OK switching off parallelism and not pursuing the second approach was: if we are going to run `eslint` only for the files modified in the branch, I guess they are a handful, not a lot. Parallelism, then, is not going to accelerate things that much and, in return, we have a clearer and more understandable solution (without that logic in the script). My assumptions about the size and number of files by PR may be totally wrong as I am still not very familiar with the common practices around here - and so the gains of parallelization in this specific case could be bigger than I thought. What do you think?

File: circle.yml (hunk 2 above)
Comment (nssw) [inline circle.yml]: Exactly, that's the purpose of this PR.

> We still want to show problems in other files/lines, but just relegate errors to warnings.

I am not sure how to integrate this idea at a CircleCI level.

#### Rationale for the approach I have chosen with this PR

The idea makes a lot of sense in our development machines, as it gives us flexibility: you can solve `warnings` on areas aside of you task or you can focus of sending good code that passes validation without being prevented to move forward due to problems in other areas. That's the developer's call (taking into account the time you have, urgency of the task, etc).

But, as far as I have understood how CircleCI is being used now in calypso, it gives us a binary answer: _does the PR pass the quality checks or not? **Is it red or green?**_. Facing that choice, I asked myself: _should CircleCI fail or should pass for `warnings` in files the PR has not modified?_ I have came to the conclusion that it should not fail for the `warnings` in those files, as the reason we have this task in the first place is being able to move forward and produce new quality code meanwhile we migrate the old code to current practices. The current [run-lint script](https://github.com/Automattic/wp-calypso/blob/master/bin/run-lint#L4) adopts the same position by reporting only `errors` to CircleCI.

#### Conclusion

So: if we are going to report only as `errors` the rules we break in the files modified in the PR **and** only those `errors` would make CircleCI to fail, I think passing `run-lint` only the files modified in the PR is a sensible thing to do. From a CircleCI point of view, we do not need to compute the files we are not going to consider as errors. For free, we also have a potential gain in performance in doing so. 

Am I missing something?

Comment (nssw) [inline circle.yml]: Exactly, that's the purpose of this PR.

> We still want to show problems in other files/lines, but just relegate errors to warnings.

I am not sure how to integrate this idea at a CircleCI level.

#### Rationale for the approach I have chosen with this PR

The idea makes a lot of sense in our development machines, as it gives us flexibility: you can solve `warnings` on areas aside of you task or you can focus of sending good code that passes validation without being prevented to move forward due to problems in other areas. That's the developer's call (taking into account the time you have, urgency of the task, etc).

But, as far as I have understood how CircleCI is being used now in calypso, it gives us a binary answer: _does the PR pass the quality checks or not? **Is it red or green?**_. Facing that choice, I asked myself: _should CircleCI fail or should pass for `warnings` in files the PR has not modified?_ I have came to the conclusion that it should not fail for the `warnings` in those files, as the reason we have this task in the first place is being able to move forward and produce new quality code meanwhile we migrate the old code to current practices. The current [run-lint script](https://github.com/Automattic/wp-calypso/blob/master/bin/run-lint#L4) adopts the same position by reporting only `errors` to CircleCI.

#### Conclusion

So: if we are going to report only as `errors` the rules we break in the files modified in the PR **and** only those `errors` would make CircleCI to fail, I think passing `run-lint` only the files modified in the PR is a sensible thing to do. From a CircleCI point of view, we do not need to compute the files we are not going to consider as errors. For free, we also have a potential gain in performance in doing so. 

Am I missing something?

File: package.json (hunk 3 above)
Comment (nssw) [inline package.json]: To make `npm run lint` run as before despite the changes I have made to `run-lint` script.  See image of changes to run-lint script.

As the purpose of this PR is to run `eslint` only for files modified within the branch (see rationale in the comment above), what happens when the branch does not modify any JS/JSX file and the result of the git expression is void? Current `run-lint` take that as an indication that it should run for the whole codebase. But, what we actually want in this specific case (at least in CircleCI) is to prevent eslint from running.

File: circle.yml (hunk 2 above)
Comment (nb) [inline circle.yml]: What if there are *errors* in the other files? What if we changed the `eslint` configuration or we upgraded `eslint` or in some other way modified how the linter works?

Comment (nb) [inline circle.yml]: > I am not sure how to integrate this idea at a CircleCI level.

I am sure you can find a way :)

File: package.json (hunk 3 above)
Comment (nb) [inline package.json]: Thanks, given your general approach this change makes sense.

File: circle.yml (hunk 1 above)
Comment (blowery) [inline circle.yml]: I think that's fine, though how does this work for the master merges?

**Comment by nylen** [comment]
> if we set to `errors` eslint rules which are now `warnings`, CircleCI report would only give contextual information about the PR, not the whole errors in the codebase.

I'm not sure about this part.  It will put the burden on PR authors to ensure that any files they modify are entirely compliant with the latest Calypso lint config, which will mean rewriting a lot of irrelevant code.

**Comment by nb** [comment]
@nylen the main idea is that the next step of the project is relegating the errors on _lines_ yo didn’t modify to warnings.

**Comment by nylen** [comment]
That would be preferable, but seems beyond the capabilities of the standard version of `eslint` as I understand them.

**Comment by nb** [comment]
@nylen definitely, @nssw started working on it in another repo, we might bring it in in Calypso: https://github.com/nssw/eslines

File: circle.yml (hunk 1 above)
Comment (nssw) [inline circle.yml]: The run-lint script is not going to run eslint.

As the HEAD in the master branch is pointing to the same place than origin/master, the result of the git expression would be void. Ergo, the run-lint script would not run.

But your question makes me wonder: I am assuming calypso is using CircleCI to do quality-checks on branches/PR before they are integrated into master. Is it useful outside that? I mean, do you expect CircleCI to also run on master?

**Comment by nssw** [comment]
> http://github.com/nssw/eslines

I was about to mention that :) The independent parts (differ to take lines from git-diff + downgrading ESLint errors to warnings) are mostly working but I need to work on putting them together and some minor things. As @nb noted the place for this is probably calypso repo, so I need also figure out a way to integrate it while preserving testing, etc.

cc @nylen

File: circle.yml (hunk 1 above)
Comment (nb) [inline circle.yml]: Yes, it’s important to run on `master`, too – you never know what will happen after the merge. Also, need to make sure we run in on direct commits to `master` (I think we have that disabled, but still, better safe than sorry).

File: circle.yml (hunk 2 above)
Comment (nssw) [inline circle.yml]: > What if there are errors in the other files?

You mean JS errors, not rule-breaking, right? This approach moves us from linting the whole codebase each time to lint the specific diff's we want to merge. Granted that all changes go through this cycle of building first a PR, we should not have undetected errors.

One concern I have after reading @blowery question above about master merges, though, is that we might need also to execute run-lint script on that cases. That is something we can do to cover all "entry points" to the codebase - if CircleCI also run on master commits. I am on mobile and I cannot check that. Would be great to learn more about how calypso uses CircleCI: does it send reports on error to some places (slack, email, etc)?

>what if we changed the eslint configuration or we upgraded ESLint or in some other way modified how the linter works?

As long as ESLint keeps accepting a list of files to check we are safe. Right now, I cannot see a way how this PR would be affected by the things you suggest - or any other changes of that kind, to that matter. I would love to be illustrated if I am missing something obvious. The point is, nevertheless, very relevant for the next step -linting only specific lines- as we are going to depend more on ESLint specifics. Thanks for the food for thought.

Comment (nssw) [inline circle.yml]: > I am sure you can find a way

Would give it a second thought.

 One approach I have already thought was: why not show the warnings too in the CircleCI console but only make CircleCI to fail (by the exit code of the script) if there are any errors in the files modified.

Reason I didn't pursue that way: current run-lint script does not show the warnings, so I took that as an indicator that people look for the green/red state given by CircleCI and only goes to its console if there are some problems.

**Comment by nssw** [comment]
This PR has been surpassed by #6945 so I am closing it. Thanks for the reviews and comments @nb @blowery and @nylen
